doc/spec/skupper-cli-spec.md

```diff
@@ -0,0 +1,287 @@
+~~~
+
+## Skupper resource commands
+
+These are the core Skupper commands.  They are not the only commands,
+however.  Additional commands will be added to the spec in the future.
+For instance, we will have commands for debugging, getting versions,
+and for installing Skupper on a platform.
+
+### `skupper <resource-type>`
+
+Print help text for the operations of this resource type.
+
+### `skupper <resource-type> create [options]`
+
+Create a resource.
+
+Resource options are set using one or more `--some-key some-value`
+command line options.  YAML resource options in camel case (`someKey`)
+are exposed as hyphenated names (`some-key`) when used as command line
+options.
+
+#### Positional arguments
+
+`token create <token-file>` - `token create` takes one positional
+parameter, the file in which to create the token.
+
+`link create <token-file>` - `link create` takes one positional
+parameter, the file containing the token.
+
+`listener create <routing-key>` - `listener create` takes one
+positional parameter, the routing key used to match the listener to
+connectors.
+
+On Kubernetes, `listener create` also requires a `--host <host>`
+option and a `--port <port>` option.
+
+`connector create <routing-key>` - `connector create` takes one
+positional parameter, the the routing key used to match the connector
+to listeners.
+
+On Kubernetes, `connector create` also requires a `--selector
+<selector>`, `--host <host>`, or `--workload <type>/<name>` option and
+a `--port` option.
+
+#### Blocking
+
+`site create` blocks until the site is ready, including ingress if
+configured.
+
+`link create` blocks until the link is active.
+
+`listener create` blocks until the router listener is ready to accept
+connections.
+
+`connector create` blocks until the router connector is ready to make
+connections.
+
+### `skupper <resource-type> delete <resource-name>`
+
+Delete a resource.
+
+Since site is a singleton, the resource name argument is not required
+for site deletion.
+
+#### Blocking
+
+The delete operation blocks until the resource is removed.
+
+### `skupper <resource-type> get [<resource-name>]`
+
+This works just like `kubectl get <type>/<name>`.  In the first
+iteration of the CLI, I think we should just delegate to kubectl.
+
+`get` without a qualifying resource name argument enumerates all the
+resources of this type.  (Same as `kubectl get`.)
+
+`get` with a resource name and `--output yaml` gives you the full
+resource YAML.  (Same as `kubectl get`.)
+
+### `skupper <resource-type> set <resource-name> [options]`
+
+Set resource options.
+
+`set` is very similar to `create`.  Instead of creating a new resource
+from defaults, it updates an existing one.
+
+This takes one or more long options (`--name foo`).  They are the same
+options as those used on create.
+
+Since site is a singleton, the resource name argument is not required
+for setting site options.
+
+#### Blocking
+
+`set` blocks with the same rules expressed for `create`.
```

Comment:
Need to consider how this will actually be implemented. Also, what happens if there are two concurrent sets? Or a set is done before a previous set completes?